### ISSUE TITLE
Error on instance property and init-only variable with the same name in a dataclass

### ISSUE THREAD

**Issue by mbraakhekke** [python/mypy#12046]
**Bug Report**

Consider the following class definition:

```python
from dataclasses import dataclass, InitVar
@dataclass
class MyClass:
    prop: InitVar[int]
    _prop: int

    def __post_init__(self, prop: int):
        self._prop = prop

    @property
    def prop(self) -> int:
        return self._prop
```
Mypy reports an error on line 10: `Name "prop" already defined on line 4`.

Is this a false positive? If not, how can I avoid this short of renaming the "`prop`" attribute or getter method (or using `#typing: ignore`)?

**Environment**
- Mypy version used: 0.931
- Python version used: 3.10.1
- Operating system and version: Windows 10 x64

**Comment by A5rocks** [comment]
This is a problem with mypy's dataclasses plugin, I suppose. However, as mypy would have already processed the class (I think?) before the plugin gets to it, I'm not certain it can be fixed. I suppose the workaround would be to `# type: ignore` the necessary lines and run mypy with `--warn-unused-ignores` (you can stick this in a config file too) to tell you when you can remove the `# type: ignore`, ie because mypy gains support.

**Comment by mbraakhekke** [comment]
Thanks. I'll implement the suggest workaround.

**Comment by joshuafishman** [comment]
I also get `"MyClass" has no attribute "prop" [attr-defined]` in mypy for any reference of that attribute -- the workaround suggested doesn't address these.

**Comment by joshuafishman** [comment]
e.g. 
```
from dataclasses import dataclass, InitVar


@dataclass
class Baz:
    foo: InitVar[int]

    @property  # type: ignore
    def foo(self) -> int:
        return 0


print(Baz(0).foo)
```
results in `error: "Baz" has no attribute "foo" [attr-defined]`

**Comment by ikonst** [comment]
Maybe if `TypeInfo` would start having a separate dict for annotations rather than put it all into `names`. Might be a bit of a refactor though, which could be less painful if `TypeInfo.names` becomes a "wrapper" around two maps, one with annotations and the other with identifiers.
